lib/handler/mruby.c

```diff
@@ -720,85 +747,106 @@ void h2o_mruby_run_fiber(h2o_mruby_generator_t *generator, mrb_value receiver, m
         if (mrb->exc != NULL)
             goto GotException;
         status = mrb_fixnum(v);
+
+        /* if no special actions were necessary, then the output is a rack response */
+        if (status >= 0) break;
+
         /* take special action depending on the status code */
-        if (status < 0) {
-            if (status == H2O_MRUBY_CALLBACK_ID_EXCEPTION_RAISED) {
-                mrb->exc = mrb_obj_ptr(mrb_ary_entry(output, 1));
-                goto GotException;
+        if (status == H2O_MRUBY_CALLBACK_ID_EXCEPTION_RAISED) {
+            mrb->exc = mrb_obj_ptr(mrb_ary_entry(output, 1));
+            generator = h2o_mruby_get_generator(mrb, mrb_ary_entry(output, 2));
+            goto GotException;
+        } else if (status == H2O_MRUBY_CALLBACK_ID_CONFIGURING_APP) {
+            mrb_value pending = mrb_ary_new_capa(mrb, 2);
+            mrb_ary_set(mrb, pending, 0, receiver);
+            mrb_ary_set(mrb, pending, 1, input);
+            mrb_ary_push(mrb, ctx->pendings, pending);
```

Comment:
PS. Or, if you would like, maybe we can have the pending list kept in the mruby-side. IIRC we have discussed the possibility before, but cannot recall why the idea was dismissed.